SHS-6191, SHS-6192, SHS-6194 : Auto-unpublish past events feature

nJim opened the pull request:
# Summary
- Adds a boolean field (machine name: field_auto_unpublish) to the event content type. This field is not required, but is a simple checkbox that will default to an off-state. 
- Moves this new field into the 'Promotion Options' group (right side 'advanced' section, see screenshot).
- Defines a new service to handle the business logic and execution of auto-unpublishing events. Defines a service to modify event node add/edit forms.
- Executes the auto unpublish on cron run. 

## Backstory
Past events are currently visible on the website, including the individual nodes and past events lists. For some sites this is useful content while others find this to be problematic. This work is part of an initiative to auto-unpublish content after an event has past.

## Need Review By (Date)
Not set.

## Urgency
Low.

## Steps to Test
1. [Log in to a test env as an admin user](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/user)
2. Setup content for testing
    1. Past event should auto unpublish. [Edit an event (this one may already be setup)](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/13821/edit) to setup this content for testing
        1. Date Range - Start: Set to a date and time in the PAST
        2. Date Range - End: Set to a date and time in the PAST
        3. Date Range - All day: Do not check
        4. Promotion options: ENABLE 'Auto-unpublish when event is past'
        5. Publish: Enabled (make sure node is published)
    2. Currently running event does not unpublish. [Edit an event (this one may already be setup)](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/13816/edit) to setup this content for testing
        1. Date Range - Start: Set to a date and time in the PAST
        2. Date Range - End: Set to a date and time in the FUTURE
        3. Date Range - All day: Do not check
        4. Promotion options: ENABLE 'Auto-unpublish when event is past'
        5. Publish: Enabled (make sure node is published)
    3. Future event does not unpublish. [Edit an event (this one may already be setup)](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/13826/edit) to setup this content for testing
        1. Date Range - Start: Set to a date and time in the FUTURE
        2. Date Range - End: Set to a date and time in the FUTURE
        3. Date Range - All day: Do not check
        4. Promotion options: ENABLE 'Auto-unpublish when event is past'
        5. Publish: Enabled (make sure node is published)
    4. Past event set to not unpublish. [Edit an event (this one may already be setup)](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/13936/edit) to setup this content for testing
        1. Date Range - Start: Set to a date and time in the PAST
        2. Date Range - End: Set to a date and time in the PAST
        3. Date Range - All day: Do not check
        4. Promotion options: DISABLE 'Auto-unpublish when event is past'
        5. Publish: Enabled (make sure node is published)
    5. All day event follows same rules. [Edit an event (this one may already be setup)](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/13931/edit) to setup this content for testing
        1. Date Range - Start: Set to a date PAST
        2. Date Range - End: Set to a date PAST
        3. Date Range - All day: ENABLED
        4. Promotion options: ENABLE 'Auto-unpublish when event is past'
        5. Publish: Enabled (make sure node is published)
3. Manually run the cron. This will take place every 12 hours on production environments.
4. [Navigate to the content overview page](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/admin/content?title=&type=hs_event&status=All) and verify that two of the test nodes have unpublished and three remain published.
    1. All day event follows same rules | Unpublished
    2. Past event should auto unpublish | Unpublished
    3. Past event set to not unpublish | Published 
    4. Future event does not unpublish | Published
    5. Currently running event does not unpublish | Published
5. Test site auto-unpublish setting:
    1. [Visit the site settings interface](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/admin/config/site-options)
    2. Set 'Auto-unpublish events once past' to 'Default: yes' and save the form.
    3. Verify that a warning message displays on form submit. This is intended to draw attention a possible misconseption about how this form work. The message informs the users that changing the sitewide setting does not update existing content.
    4. [Add a new hs_event node](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/add/hs_event). Verify that the "Promotion options" > "Auto-unpublish when event is past" defaults to enabled.
    5. [Return to the site settings interface](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/admin/config/site-options) and set 'Auto-unpublish events once past' to 'Default: no'. Save the form.
    6. [Add a new hs_event node](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/node/add/hs_event). Verify that the "Promotion options" > "Auto-unpublish when event is past" defaults to disabled.
7. Test the bulk operations:
    1. [Navigate to an event's management view](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/admin/content/manage/event) to explore the new node bulk operations for events.
    2. Identify one or more event nodes. Edit these nodes in a new tab to verify that "Auto-unpublish when event is past" is NOT enabled. 
    3. Use the bulk "Modify field values" operation to set "Auto-unpublish when event is past" to"On" for these nodes. Manually edit the nodes to verify that the bulk operation is able to enable this field.
    4. [Visit the views display for another content bundle](https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com/admin/structure/views/view/hs_default_manage_content/edit/people) to verify that this bulk action does not exist for other content types.

# Review Tasks

## Backend / Functional Validation
### Code
- [ ] Are the naming conventions following our standards?
- [ ] Are PHP functions and variables in `snake_case` and not `camelCase`?
- [ ] Does Drupal code follow [Drupal Coding Standards](https://www.drupal.org/docs/develop/standards/php/php-coding-standards)?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided?

### Code security
- [ ] Is all [user input properly sanitized when rendered](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

### General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211009492118978
  - https://app.asana.com/0/0/1210850524973556

nJim committed:
feat(SHS-6191): add auto unpublish field to event content type

nJim committed:
feat(SHS-6191): add auto unpublish service to run on cron

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-colorful-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-traditional-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://english-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://history-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://west-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959

Visual Diffs:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959/visualdiffs

nJim committed:
feat(SHS-6191): move auto unpublish field to active space on node form

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-colorful-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-traditional-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://english-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://history-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://west-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959

Visual Diffs:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959/visualdiffs

nJim committed:
feat(SHS-6192): set default auto unpublish config

nJim committed:
feat(SHS-6194): define auto unpublish setting bulk action

nJim committed:
feat(SHS-6194): add bulk action to manage events interface

nJim committed:
SHS-6191: Add auto-unpublish functionality for events

- Add EnableAutoUnpublish action plugin
- Add field_auto_unpublish field to hs_event content type
- Add system action configuration
- Add README documentation
- Clean up duplicate configuration files

nJim committed:
SHS-6191: Clean up duplicate configuration files

nJim committed:
Merge branch 'SHS-6191-auto-unpublish' of github.com:SU-HSDO/suhumsci into SHS-6191-auto-unpublish

nJim committed:
fix(SHS-6194): remove readme

nJim committed:
feat(SHS-6192): set default auto unpublish config

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-colorful-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-traditional-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://english-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://history-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://west-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959

Visual Diffs:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959/visualdiffs

Jnwibowo7321 reviewed (approved):
Tested on tugboat and it LGTM!

Jnwibowo7321 commented:
This is ready for review by @joegl, Is that correct @nJim @cienvaras ?

nJim commented:
> This is ready for review by @joegl, Is that correct @nJim @cienvaras ?

This still needs our internal review. I believe Andrés will review this early next week.

joegl commented:
Please re-add me for review once it's ready, thanks 👍

cienvaras committed:
Merge branch '11.12.3-release' of github.com:SU-HSDO/suhumsci into SHS-6191-auto-unpublish

cienvaras commented on docroot/profiles/humsci/su_humsci_profile/su_humsci_profile.install:
```diff
@@ -1553,3 +1553,24 @@ function su_humsci_profile_update_9736() {
+ * Add field_auto_unpublish field to bottom of the event form display.
+ */
+function su_humsci_profile_update_9737() {
+  $entity_type_manager = \Drupal::entityTypeManager();
```
Please see [this feedback](https://github.com/SU-HSDO/suhumsci/pull/1897#pullrequestreview-3064341307) left by SWS in the Editable Metatags PR. This case is very similar, so it's probably necessary to import the field before adding it to the form display here too.

nJim replied:
@cienvaras I've updated the hook to follow the same pattern for adding the field within updb.

cienvaras reviewed (changes requested):
@nJim Looks great, everything works as expected. Just one possible fix is needed, please check the comment in the code.

Also, I wasn't able to test the bulk operation in the views admin interface preview. I had to enable the default view and disable the custom one in order to test it. This is unrelated with your work on the PR, but maybe it's a good idea to update the testing instructions for the H&S team.

nJim committed:
Merge branch '11.13.1-release' into SHS-6191-auto-unpublish

nJim committed:
fix(SHS-6191): place unpublish field with update hook

cienvaras reviewed (approved):
@nJim Thanks for the updates, everything works as expected!

@Jnwibowo7321 @ahughes3 Ready for you

joegl commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    if ($unpublished_count > 0) {
+      $logger->info('Auto-unpublished @count past events during cron run.', [
```
I'd take "during cron run" out of this. The service itself isn't specifically triggered by cron and might be run outside of cron at some point.

joegl commented on config/default/system.action.hs_events_enable_auto_unpublish.yml:
```diff
@@ -0,0 +1,12 @@
+uuid: 92f31f39-5fb9-40f6-9c18-1fd43f6e6f24
```
Is a custom action for VBO necessary? Wouldn't using the default set field value action suffice for setting this value on event nodes in bulk?

nJim replied:
The team discussed and agreed. We will remove this action in favor of the field value one that already exists.

joegl commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php:
```diff
@@ -0,0 +1,203 @@
+  /**
+   * Node type for events.
+   */
+  const EVENT_NODE_TYPE = 'hs_event';
```
I really like the use of constants in this class 👍

joegl commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php:
```diff
@@ -0,0 +1,203 @@
+   * @param array &$form
+   *   The form array to modify.
+   */
+  public function setAutoUnpublishDefaultValue(array &$form): void {
```
These two form methods might make sense as their own service instead of lumping into the AutoUnpublish method. Separation of concerns. Up to you though 👍

joegl commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php:
```diff
@@ -0,0 +1,203 @@
+/**
+ * Service for auto-unpublishing past events and handling form modifications.
+ */
+class AutoUnpublish {
```
Even though this lives in the `hs_events` module I would still considering renaming the service to `AutoPublishEvents` to make it more explicit in what its intention is.

joegl commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php:
```diff
@@ -0,0 +1,203 @@
+    if (empty($nids)) {
+      return 0;
+    }
+
```
Should there be some type of batching here? Either adding a range to the query or only loading a specific item of node id's from storage at a time? I don't think it's absolutely necessary, but it couldn't hurt especially if this gets triggered on a lot of nodes on the first run.

nJim replied:
I'm adding a default limit to this method set to 100 items. 

The batch API is generally not the right fit for something executed by cron. Creating a queue worker is a better patterns, but may create a delay in execution. So keeping it simple seems to be the best route.

Jnwibowo7321 commented:
LGTM on my end on Tugboat. Ready for your review @joegl

joegl commented on docroot/modules/humsci/hs_events/hs_events.module:
```diff
@@ -19,3 +20,29 @@ function hs_events_hs_entity_view_alter(array &$build, HsEntityInterface $entity
+ * Implements hook_form_FORM_ID_alter().
+ */
+function hs_events_form_node_hs_event_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  $auto_unpublish_service = \Drupal::service('hs_events.auto_unpublish');
```
Can you add a comment above to describe what this is manipulating?

joegl commented on docroot/modules/humsci/hs_events/hs_events.module:
```diff
@@ -19,3 +20,29 @@ function hs_events_hs_entity_view_alter(array &$build, HsEntityInterface $entity
+ * Implements hook_form_FORM_ID_alter().
+ */
+function hs_events_form_node_hs_event_edit_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  $auto_unpublish_service = \Drupal::service('hs_events.auto_unpublish');
```
Can you add a comment above to describe what this is manipulating?

joegl commented on docroot/modules/humsci/hs_events/hs_events.module:
```diff
@@ -19,3 +20,29 @@ function hs_events_hs_entity_view_alter(array &$build, HsEntityInterface $entity
+ */
+function hs_events_cron() {
+  // Use the auto-unpublish service to handle past events.
+  $auto_unpublish_service = \Drupal::service('hs_events.auto_unpublish');
```
Can you add a comment above to describe what this cron run is doing?

joegl commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php:
```diff
@@ -0,0 +1,203 @@
+        $unpublished_count++;
+
+        // Log the action for debugging purposes.
+        $logger->notice('Auto-unpublished event "@title" (ID: @id) as it is past its end date.', [
```
I think "info" might be more appropriate here than "notice".

joegl reviewed (changes requested):
Looks great. Very minimal feedback and suggestions.

Jnwibowo7321 commented:
Since Jim is out this whole week. Is this something that you can resolve to make this in this week @cienvaras? 
Please re-assign this to you if you think it is doable, if not we can wait for Jim.

nJim committed:
fix(SHS-6191): refactor based on PR feedback

nJim committed:
fix(SHS-6191): remove custom views bulk action

nJim committed:
feat(SHS-6191): add validation message when changing publishing settings

nJim committed:
feat(SHS-6191): merge release

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1921-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-colorful-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://hs-traditional-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://english-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://history-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com
- https://west-qtsbiyzxpq2aftkdchcfqbpjwkvutdes.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959

Visual Diffs:
- https://dashboard.tugboatqa.com/6887a555307b0ac27d854959/visualdiffs

nJim commented on docroot/modules/humsci/hs_events/src/Services/AutoUnpublishEvents.php:
```diff
@@ -0,0 +1,127 @@
+   * @return int
+   *   The number of events that were unpublished.
+   */
+  public function unpublishPastEvents(int $limit = 100): int {
```
Good note on large sets of nodes. Setting a default limit to avoid long running tasks. It will have to run multiple times for sites that are trying to update a large number of nodes.

If needed, we can use a version of this method in the future for something like a 'batch update now' button that a user can run from the admin interfaces. But I avoid trying to do batches in cron jobs.

joegl replied:
No batches in Cron jobs makes sense. The limit is a good compromise. Using the Ultimate Cron page should also enable Administrators to run this Cron task manually if necessary.

nJim commented on docroot/modules/humsci/hs_events/src/Services/EventFormModifier.php:
```diff
@@ -0,0 +1,128 @@
+/**
+ * Service for modifying event forms and handling form auto-unpublish logic.
+ */
+class EventFormModifier {
```
Separation of concerns. Moved form methods to this class.

nJim commented on docroot/modules/humsci/hs_events/hs_events.module:
```diff
@@ -19,3 +20,74 @@ function hs_events_hs_entity_view_alter(array &$build, HsEntityInterface $entity
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function hs_events_form_config_pages_hs_site_options_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  // Add validation handler to check for auto-unpublish field changes.
+  $form['#validate'][] = 'hs_events_site_options_form_validate';
+
+  // Store original values for comparison on validation.
+  if (!$form_state->has('original_values')) {
+    $original_values = [];
+    if (isset($form['field_site_auto_unpublish']['widget']["#default_value"][0])) {
+      $original_values['field_site_auto_unpublish'] = $form['field_site_auto_unpublish']['widget']["#default_value"][0];
+    }
+    $form_state->set('original_values', $original_values);
+  }
+}
+
+/**
+ * Validation handler for the site options form.
+ *
+ * Checks if the auto-unpublish field value has changed and displays a warning
+ * message to inform users that changing the default setting does not affect
+ * existing events.
+ *
+ * @param array $form
+ *   The form array.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state object.
+ */
+function hs_events_site_options_form_validate($form, FormStateInterface $form_state) {
+  // Check if the auto-unpublish field value has changed.
+  $original_value = $form_state->get('original_values')['field_site_auto_unpublish'] ?? NULL;
+  $new_value = $form_state->getValue(['field_site_auto_unpublish', 0, 'value']);
+
+  if ($original_value !== $new_value) {
+    $message = t('Changing the default for "Auto-unpublish events once past" does not impact existing events. Visit <a href="/admin/content/manage/event">the list of all events</a>, select the events that you want to change, and use the "Modify field values" bulk action to set the "Auto-unpublish when event is past" field to "On" or "Off".');
+    \Drupal::messenger()->addWarning($message);
+  }
+}
```
This code is new since last PR review. Based on your suggestion, we removed the custom Views Bulk Action and instead will train folks to use the existing 'Modify field values' action. 

We are also afraid that folks will change this global setting and may not realize it does not retroactively change existing nodes. So an additional warning message is set if someone changes this site-wide settings. A validation handler had to be used as the "config_pages" modules is not setup to allow custom submit handlers.

nJim commented:
@joegl This PR is once again ready for review. I've added comments in GitHub to highlight any new code. Changes includes:
- New comments, renaming a class, and other small changes to align with your project coding standards.
- Breaking the original service into two services (on for unpublishing and one for altering forms) to better separate concerns.
- Removing the custom views bulk action. Adding a instructions via a custom form validation to explain how the 'Modify field values' action can be used in this capacity.
- Setting a limit of 100 nodes to update per cron job.

joegl committed:
Merge branch '11.14.1-release' into SHS-6191-auto-unpublish

joegl committed:
Clean-up.

joegl reviewed (approved):
This is great @nJim, thanks! 👍
